Geometry_Engine/Modify/Offset.cs

```diff
@@ -0,0 +1,868 @@
+        {
+            if (offset == 0)
+                return curve.Clone();
+
+            double radius = curve.Radius;
+            Circle result = curve.Clone();
+
+            if (normal == null)
+                normal = curve.Normal();
+
+            if (normal.DotProduct(curve.Normal()) > 0)
+                radius += offset;
+            else
+                radius -= offset;
+
+            if (radius > 0)
+            {
+                result.Radius = radius;
+                return result;
+            }
+            else
+            {
+                Reflection.Compute.RecordError("Offset value is greater than circle radius");
+                return null;
+            }
+        }
+
+        /***************************************************/
+
+        [NotImplemented]
+        public static Ellipse Offset(this Ellipse curve, double offset, Vector normal = null, bool tangentExtensions = false, double tolerance = Tolerance.Distance)
+        {
+            throw new NotImplementedException();
+        }
+
+        /***************************************************/
+
+        [NotImplemented]
+        public static NurbsCurve Offset(this NurbsCurve curve, double offset, Vector normal = null, bool tangentExtensions = false, double tolerance = Tolerance.Distance)
+        {
+            throw new NotImplementedException();
+        }
+
+        /***************************************************/
+
+        [Description("Creates an offset of a curve. Works only on planar curves")]
+        [Input("curve", "Curve to offset")]
+        [Input("distance", "Offset distance. Positive value offsets outside of a curve. If normal is given then offsets to the right with normal pointing up and direction of a curve pointing forward")]
+        [Input("tangentExtensions", "If true, arc segments of a PolyCurve will be extend by a tangent line, if false - by arc")]
+        [Input("normal", "Normal of a plane for offset operation, not needed for closed curves")]
+        [Output("curve", "Resulting offset")]
+        public static Polyline Offset(this Polyline curve, double offset, Vector normal = null, bool tangentExtensions = false, double tolerance = Tolerance.Distance)
+        {
+            if (!curve.IsPlanar())
+            {
+                BH.Engine.Reflection.Compute.RecordError("Offset works only on planar curves");
+                return null;
+            }
+            bool isClosed = curve.IsClosed(tolerance);
+            if (normal == null)
+                if (!isClosed)
+                    BH.Engine.Reflection.Compute.RecordError("Normal is missing. Normal vector is not needed only for closed curves");
+                else
+                    normal = curve.Normal();
+            else
+                normal = normal.Normalise();
+
+            if (offset == 0)
+                return curve.Clone();
+
+            if (offset > 0.05 * curve.Length())
```

Comment:
Where does 0.05 come from? 🐒

Reply:
Method first offset each SubCurve and then fillet them. For greater offset values it is hard to recreate original shape of a curve and that is why I decided to in those cases split it into smaller offsets. 0.05 * Length is a value I came up with experimentally and seems to work best. As said in additional comments, I will later raise another issue in which I'll point what else is wrong and could be improved.

Reply:
Of course we can get rid of this loop but it would reduce effectiveness of whole method.. At least in some cases.

Reply:
It's working well under testing, but I might be tempted to expose it as an input parameter either now (if easy) or later (if more than 30 seconds work) due to the requirement of this PR.